### PR TITLE
feat: add warning icon for unreconciled Purchase Invoices in Payment Entry references

### PR DESCRIPTION
**Pull request by divyam-mistry**
image

image

### CODE REVIEW

File: india_compliance/gst_india/overrides/payment_entry.py [hunk 1]
```diff
@@ -14,6 +16,40 @@
...
 
+@frappe.whitelist()
+def get_reconciliation_status_for_invoice_list(invoice_list):
+    if not invoice_list:
+        return
+
+    if isinstance(invoice_list, str):
+        invoice_list = json.loads(invoice_list)
+
+    original_list = frappe.db.get_list(
```
Comment (sagarvora) [inline india_compliance/gst_india/overrides/payment_entry.py]: Use `frappe.get_list`. Note: `get_list` checks permissions implicitly, if you intend to not check those, you must specify the reason and then use `frappe.get_all` instead.

File: india_compliance/gst_india/overrides/payment_entry.py [hunk 2]
```diff
@@ -14,6 +16,40 @@
...
+        "Purchase Invoice",
+        filters={"name": ["in", invoice_list]},
+        fields=("name", "reconciliation_status"),
+    )
+    transformed_dict = {
+        item["name"]: item["reconciliation_status"] for item in original_list
+    }
```
Comment (sagarvora) [inline india_compliance/gst_india/overrides/payment_entry.py]: You don't need dict comprehension. This should do the trick:

`dict(frappe.get_list(..., ..., as_list=True))`

File: india_compliance/gst_india/overrides/payment_entry.py [hunk 3]
```diff
@@ -14,6 +16,40 @@
 from india_compliance.gst_india.utils import get_all_gst_accounts
 
 
+@frappe.whitelist()
+def get_reconciliation_status_for_invoice_list(invoice_list):
+    if not invoice_list:
+        return
+
+    if isinstance(invoice_list, str):
+        invoice_list = json.loads(invoice_list)
```
Comment (sagarvora) [inline india_compliance/gst_india/overrides/payment_entry.py]: Use `frappe.parse_json`

File: india_compliance/gst_india/client_scripts/payment_entry.js [hunk 4]
```diff
@@ -1,51 +1,148 @@
...
-		update_gst_details(
-			frm,
-			"india_compliance.gst_india.overrides.payment_entry.update_party_details"
-		);
-	},
-
-	customer_address(frm) {
-		update_gst_details(frm);
-	},
+
```
Comment (sagarvora) [inline india_compliance/gst_india/client_scripts/payment_entry.js]: why newline here?

**Review by sagarvora — CHANGES_REQUESTED** [review]

File: india_compliance/gst_india/client_scripts/payment_entry.js [hunk 5]
```diff
@@ -1,51 +1,147 @@
...
+    party(frm) {
+        update_gst_details(
+            frm,
+            "india_compliance.gst_india.overrides.payment_entry.update_party_details"
+        );
+    },
+
+    customer_address(frm) {
+        update_gst_details(frm);
+    },
```
Comment (vorasmit) [inline india_compliance/gst_india/client_scripts/payment_entry.js]:
```suggestion
    customer_address: update_gst_details,
```

File: india_compliance/gst_india/client_scripts/payment_entry.js [hunk 6]
```diff
@@ -1,51 +1,147 @@
...
 
+function override_get_outstanding_documents(frm) {
+    const old_fn = frm.events?.get_outstanding_documents;
+    if (!old_fn) return;
+
+    const new_fn = function () {
+        old_fn(...arguments);
+        frappe.after_ajax(() => {
+
+            const invoice_list = frappe.last_response && frappe.last_response.message
```
Comment (vorasmit) [inline india_compliance/gst_india/client_scripts/payment_entry.js]: How about using this with `OR`
https://javascript.info/optional-chaining

File: india_compliance/gst_india/client_scripts/payment_entry.js [hunk 7]
```diff
@@ -1,51 +1,147 @@
...
+                ? frappe.last_response.message
+                    .filter(r => r.voucher_type === "Purchase Invoice")
+                    .map(r => r.voucher_no)
+                : [];
+
+            frappe.call({
+                method: "india_compliance.gst_india.overrides.payment_entry.get_reconciliation_status_for_invoice_list",
```
Comment (vorasmit) [inline india_compliance/gst_india/client_scripts/payment_entry.js]: How about using `override_whitelisted_method`? This is purely frontend functionality and should not break the backend.

File: india_compliance/gst_india/client_scripts/payment_entry.js [hunk 8]
```diff
@@ -1,51 +1,147 @@
...
+    if (!reconciliation_status_dict) return;
+
+    let rows = frm.fields_dict.references.grid.grid_rows
+        .filter(r => r.doc.reference_doctype === "Purchase Invoice");
+
+    if (name) rows = rows.filter(r => r.doc.reference_name === name);
+
+    for (const row of rows) {
+
+        if (row.doc.reference_name in reconciliation_status_dict && reconciliation_status_dict[row.doc.reference_name] !== 'Unreconciled') continue;
```
Comment (vorasmit) [inline india_compliance/gst_india/client_scripts/payment_entry.js]: What should happen if row.reference_name is not in reconciliation_status_dict

File: india_compliance/gst_india/client_scripts/payment_entry.js [hunk 9]
```diff
@@ -1,51 +1,147 @@
...
+        const target_div = row.columns.reference_name;
+        const is_warning_icon_already_present = $(target_div).find(".warning-icon").length > 0;
+
+        if (is_warning_icon_already_present) continue;
+
+        $(WARNING_ICON).appendTo(target_div);
+
+        $('.warning-icon').hover(
+            function () {
+                $(this).attr('title', '2A/2B Status: Unreconciled');
```
Comment (vorasmit) [inline india_compliance/gst_india/client_scripts/payment_entry.js]: does it matter if the title is always part of HTML, even without hover?

Comment (divyam-mistry) [reply india_compliance/gst_india/client_scripts/payment_entry.js]: Yes, this is not necessary. I've now added it in HTML itself.

File: india_compliance/gst_india/client_scripts/payment_entry.js (hunk 8 above)
Comment (divyam-mistry) [reply india_compliance/gst_india/client_scripts/payment_entry.js]: Will this change suffice?

```
if (!reconciliation_status_dict[row.doc.reference_name] || reconciliation_status_dict[row.doc.reference_name] !== 'Unreconciled') continue;
```

**Pull request merged**